server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -317,14 +327,96 @@ public IndexInput openInput(String name, IOContext context) throws IOException {
         }
     }
 
-    public void copyFrom(Directory from, String src, String dest, IOContext context, boolean useCommonSuffix) throws IOException {
+    public void copyFilesFrom(Directory from, Collection<String> files, IOContext context) throws Exception {
+
+        List<CompletableFuture<UploadResponse>> resultFutures = new ArrayList<>();
+
+        for (String src : files) {
+            String remoteFilename = createRemoteFileName(src, false);
+            if (remoteDataDirectory.getBlobContainer().isMultiStreamUploadSupported()) {
+                CompletableFuture<UploadResponse> resultFuture = createUploadFuture(from, src, remoteFilename, context);
+                resultFutures.add(resultFuture);
+            } else {
+                copyFrom(from, src, src, context, false);
+            }
+        }
+
+        if (resultFutures.isEmpty() == false) {
+            CompletableFuture<Void> resultFuture = CompletableFuture.allOf(resultFutures.toArray(new CompletableFuture[0]));
+            try {
+                resultFuture.get();
```

Comment:
In case of an error, exception is thrown which is propagated to `resultFuture`. `resultFuture` then calls the `listener.onFailure` to notify the same. Now, if there's a failure in future itself rather than upload then the outer exception block will handle it and call listener.